src/query/xasl_cache.c

```diff
@@ -1443,7 +1459,10 @@ xcache_insert (THREAD_ENTRY * thread_p, const COMPILE_CONTEXT * context, XASL_ST
     }
   else
     {
-      if (xcache_Soft_capacity < xcache_Entry_count && xcache_Cleanup_flag == 0)
+      gettimeofday (&current_time, NULL);
+      if ((xcache_Soft_capacity < xcache_Entry_count
+	   || TIME_DIFF_SEC (current_time, xcache_last_cleaned_time) > xcache_time_threshold)
```

Comment:
I'd like to propose to introduce a function something like this.

```c
typedef enum xcache_cleanup_reason XCACHE_CLEANUP_REASON;
enum xcache_cleanup_reason
{
  XCACHE_CLEANUP_NONE; /* no cleanup is required */
  XCACHE_CLEANUP_FULL;
  XCACHE_CLEANUP_TIMEOUT;
};

static XCACHE_CLEANUP_REASON
xcache_need_cleanup (void)
{
  if (xcache_Soft_capacity < xcache_Entry_count)
    {
      return XCACHE_CLEANUP_FULL;
    }
  else
    {
      ...
      if (TIME_DIFF_SET (current_time, xcache_last_cleaned_time) > xcache_time_threshold)
        {
          return XCACHE_CLEANUP_TIMEOUT;
        }
    }

  return XCACHE_CLEANUP_NONE;
}
```

You may think of better names for enums and functions :)

Then, you can rewrite `if` with the new function. 
```c
  if (xcache_need_cleanup () != XCACHE_CLEANUP_NONE
      && xcache_Cleanup_flag == 0)
```